src/EditorFeatures/Core/Implementation/IntelliSense/Helpers.cs

```diff
@@ -1,27 +1,85 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor.GoToDefinition;
+using Microsoft.CodeAnalysis.Editor.Host;
 using Microsoft.VisualStudio.Text.Adornments;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense
 {
     internal static class Helpers
     {
-        internal static IEnumerable<object> BuildClassifiedTextElements(ImmutableArray<TaggedText> taggedTexts)
+        internal static IEnumerable<object> BuildClassifiedTextElements(ImmutableArray<TaggedText> taggedTexts, Document document, IEnumerable<Lazy<IStreamingFindUsagesPresenter>> streamingPresenters)
+        {
+            var index = 0;
+            return BuildClassifiedTextElements(taggedTexts, ref index, document, streamingPresenters);
+        }
+
+        private static IReadOnlyCollection<object> BuildClassifiedTextElements(ImmutableArray<TaggedText> taggedTexts, ref int index, Document document, IEnumerable<Lazy<IStreamingFindUsagesPresenter>> streamingPresenters)
         {
             // This method produces a sequence of zero or more paragraphs
             var paragraphs = new List<object>();
 
             // Each paragraph is constructed from one or more lines
-            var currentParagraph = new List<ClassifiedTextElement>();
+            var currentParagraph = new List<object>();
 
             // Each line is constructed from one or more inline elements
             var currentRuns = new List<ClassifiedTextRun>();
 
-            foreach (var part in taggedTexts)
+            for (; index < taggedTexts.Length; index++)
```

Comment:
fwiw i find loops that increment both on hte outside, and on the inside hard to reason about.  Would you be willing to have this be a `while` and just be more explicit inside about movin the index forward the right amount?